ddtrace/contrib/internal/vertexai/_utils.py

```diff
@@ -0,0 +1,302 @@
+    if not model_name or not isinstance(model_name, str):
+        return ""
+    if "/" in model_name:
+        return model_name.split("/")[-1]
+    return model_name
+
+def get_system_instruction_parts_from_model(instance):
+    """
+    Assumes that the system instruction is provided as []Part
+    """
+    return getattr(instance, "_system_instruction", None)
+
+def get_generation_config_from_model(instance, kwargs):
+    """
+    The generation config can be defined on the model instance or 
+    as a kwarg of the request. Therefore, try to extract this information
+    from the kwargs and otherwise default to checking the model instance attribute.
+    """
+    generation_config_arg = kwargs.get("generation_config", {})
+    if generation_config_arg != {}:
+        return generation_config_arg if isinstance(generation_config_arg, dict) else generation_config_arg.to_dict()
+    generation_config_attr = instance._generation_config or {}
+    return generation_config_attr if isinstance(generation_config_attr, dict) else generation_config_attr.to_dict()
+
+def extract_info_from_parts(parts):
+    """Return concatenated text from parts and function calls."""
+    concatenated_text = ""
+    function_calls = []
+    for part in parts:
+        text = getattr(part, "text", "")
+        if text:
+            concatenated_text += text
+        function_call = getattr(part, "function_call", None)
+        if function_call is not None:
+            function_calls.append(function_call)
+    return concatenated_text, function_calls
+
+def _tag_response_parts(span, integration, parts):
+    text, function_calls = extract_info_from_parts(parts)
+    span.set_tag_str(
+        "vertexai.response.candidates.%d.content.parts.%d.text" % (0, 0),
+        integration.trunc(str(text)),
+    )
+    for idx, function_call in enumerate(function_calls):
+        span.set_tag_str(
+            "vertexai.response.candidates.%d.content.parts.%d.function_calls.%d.function_call.name" % (0, 0, idx),
+            integration.trunc(str(getattr(function_call, "name", ""))),
+        )
+        span.set_tag_str(
+            "vertexai.response.candidates.%d.content.parts.%d.function_calls.%d.function_call.args" % (0, 0, idx),
+            integration.trunc(str(getattr(function_call, "args", ""))),
+        )
+
+def tag_stream_response(span, chunks, integration):
+    all_parts = []
+    role = ""
+    for chunk in chunks:
+        for candidate_idx, candidate in enumerate(chunk.candidates):
+            finish_reason = candidate.finish_reason
+            if finish_reason:
+                span.set_tag_str(
+                    "vertexai.response.candidates.%d.finish_reason" % (candidate_idx), str(finish_reason.name)
+                )
+            candidate_content = candidate.content
+            role = role or candidate_content.role
+            span.set_tag_str("vertexai.response.candidates.%d.content.role" % (candidate_idx), str(role))
```

Comment:
are we tagging role and finish reason for every single chunk?

Reply:
Might be worth constructing the whole response and then tagging role/finish_reason at the end instead of intermittently for every chunk (this results in unnecessary operations)

Reply:
finish_reason will only be included in the last chunk. As for the role, there is only role that needs to be tagged, so I'm going to move it out of the for loop :)